### PR TITLE
fix(quill): stop heatmap brush edges snapping into neighboring rows

### PR DESCRIPTION
**Pull request by jonmcwest**
## Problem

The chart core's 2D brush (`onAreaSelect` + the Heatmap's `onBrush`) shipped to master as part of the tracing brush wiring ([#71576](https://github.com/PostHog/posthog/pull/71576) carried the whole stack when it merged). What master got was an earlier snapshot of this branch — this PR now holds only the review follow-ups that landed here afterwards.

One of them is a real behavior fix: a brush edge released exactly on a row boundary floors into the neighboring row, so the reported row range includes a row the rectangle only touches.

## Changes

- **Fix — boundary over-selection:** inset each brush edge by half a pixel before snapping it to a row (`BRUSH_EDGE_EPSILON_PX`), so an edge on a row boundary resolves to the row the rectangle actually covers. The `FULL_HEIGHT_MIN_Y_DRAG_PX` guard keeps the span far thicker than the nudge, so it can't invert.
- **Refactor:** `AreaSelectData` and `DateRangeZoomData` now share a `LabelRange` base type instead of duplicating the label/index fields.
- **Tidy-ups from review:** `rawRowAtY` extracted in `heatmap-layout.ts` (clamped vs unbounded row lookup made explicit), destructuring cleanup in `canvas-renderer.ts`, and a simplified dispatch branch in `useDragToZoom.ts`.
- The `BrushSelection` story keeps master's `test-skip` tag (the brush overlay only appears mid-drag, so a static snapshot adds no coverage) — the stale visual baselines this branch previously added are dropped.

### Screenshot (Storybook, run on a devbox)

The `BrushSelection` story mid-drag — the gesture whose edge snapping this PR fixes:

Heatmap 2D brush selection

## How did you test this code?

- Full quill-charts + tracing suites under the frontend jest project after reconciling with master: 83 suites / 1840 tests passed.
- The Heatmap brush tests cover the row-range mapping via `rawDrag`; the epsilon guard keeps the existing diagonal-drag and flat-drag cases passing while fixing the boundary case.
- `pnpm --filter=@posthog/frontend typescript:check` clean.

## Automatic notifications

- [ ] Publish to changelog?
- [ ] Alert Sales and Marketing teams?

## Docs update

`packages/quill/packages/charts/AGENTS.md` already documents `onAreaSelect`/`onBrush` (landed with the earlier snapshot); no behavior documented there changes here.

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

Authored with PostHog Code (Claude). This PR originally introduced the 2D brush; after an out-of-band merge of the stacked tracing PR carried an earlier snapshot of it to master, the branch was reconciled with master so only the post-snapshot review fixes remain. The half-pixel inset was chosen over an exclusive-bound row lookup because it fixes the boundary case without changing the row math the clamped lookup shares with hover resolution.

---
*Created with [PostHog Code](https://posthog.com/code?ref=pr)*

### CODE REVIEW

**Comment by jonmcwest** [comment]
## 🧹 Polish pass (simplify + code-review)

Ran the `simplify` and `code-review` passes back-to-back on this PR's diff. Fixes are pushed to `posthog-code/quill-2d-brush` in `refactor(quill): tidy 2D brush internals after review` (f5413d5) — this comment is the summary.

### ✨ Simplify — applied

| File | Change |
| --- | --- |
| `heatmap-layout.ts` | Extracted a shared `rawRowAtY` helper so `rowAtY` and `rowAtYClamped` no longer duplicate the bottom-up row-index formula. Each keeps its own edge policy (`-1` vs clamp). |
| `core/canvas-renderer.ts` | In `composeDrawHoverWithSelection`, destructured `y0`/`y1` into narrowed locals so TypeScript sees the `hasY` guarantee — removing the dead `?? 0` fallbacks that could never fire and read as false defensiveness. |
| `core/hooks/useDragToZoom.ts` | Hoisted the shared `{ startLabel, endLabel, startIndex, endIndex }` payload in `completeDrag` so the `onAreaSelect` and `onDateRangeZoom` branches stop repeating the object shape and the `labelsRef.current` lookup. Precedence (area-select wins) is unchanged. |

### 🔍 Code-review (high effort) — applied

| File | Finding | Fix |
| --- | --- | --- |
| `core/types.ts` | `ChartDrawArgs.dragRect`'s doc still said "x-axis only", but the field now carries `y0`/`y1` for a 2D brush. Stale doc → a future adapter reading `dragRect.y0/y1` off draw args would assume it's never set. | Updated the comment to describe both the x drag-to-zoom range and the 2D vertical range. |

### 👀 Found but not applied

- **`useChartInteraction.ts` — `onAreaSelectWithScales` is `undefined` when `scales` is null.** If `scales` recomputed to `null` *mid-drag* (a resize/reflow during an in-progress brush), the completed gesture would be dropped, and — for a hypothetical consumer wiring both `onAreaSelect` **and** `onDateRangeZoom` — could fall through to the zoom callback. **Not changed** because it isn't reachable in practice: `Heatmap` exposes only `onBrush` (→ `onAreaSelect`) and no `onDateRangeZoom`, so the wrong-callback fallthrough can't occur; and with null `scales` there's no layout to map pixels onto, so dropping the gesture is the correct degradation. The "deep" fix (capture `scales` at drag-start) would couple the deliberately scale-agnostic `useDragToZoom` hook to `scales`, which the documented seam (`AGENTS.md`) avoids. Flagging for awareness only.

### ✅ Verified clean

- **Altitude** — the `scales`-threading seam (`onAreaSelectWithScales` mirroring `wrapClickData`), the `FULL_HEIGHT_MIN_Y_DRAG_PX = 8` flat-drag threshold living in the Heatmap adapter, and the `_private.__heatmap` back-channel are all at the right layer and consistent with existing chart adapters.
- **Conventions** — `AGENTS.md`/`CLAUDE.md` (a symlink to it) updated, a `BrushSelection` story added, explicit return types present, no inline styles, comments explain *why*.
- **Removed-behavior / cross-file** — the widened `zoomEnabled`/`crosshair` gates, the `completeDrag(mouseX, mouseY)` signature, and the optional `DragRect.y0/y1` don't regress any `onDateRangeZoom`-only chart (`trackY` stays false → `y0/y1` never set → full-height fallback preserved).

### 🧪 Test / lint status

This sandbox has no installed `node_modules`, so `jest`/`tsc`/`oxlint` could not be run here. All applied changes are mechanical and behavior-preserving (identical arithmetic, identical clamped values, identical callback payloads, one comment). The PR's existing coverage — the `Heatmap.test.tsx` `onBrush` suite and the `canvas-renderer.test.ts` selection-rect cases — exercises these paths and remains valid. Please rely on CI for the authoritative run.

**Comment by jonmcwest** [comment]
Rebuilt this branch to resolve the merge conflict with master.

This PR was stacked on the heatmap-chart branch, which has since merged to master. The branch history had drifted (an older copy of the heatmap chart plus unrelated snapshot-hash churn), which is what produced the conflict. I rebased the two 2D-brush commits onto current master — the brush was authored against a charts tree identical to today's master, so they cherry-picked cleanly.

Net diff is now just the 2D brush feature (12 files under `packages/quill/packages/charts/`); the stale heatmap/snapshot files that were showing up before are gone. CI is green. Only remaining gate is a required review.

<sub>🦉 via [talyn.dev](https://talyn.dev)</sub>

File: packages/quill/packages/charts/src/charts/Heatmap/heatmap-layout.ts [hunk 1]
```diff
@@ -53,10 +53,24 @@ export function rowAtY(layout: HeatmapLayout, y: number): number {
...
 }
 
+/** Like `rowAtY`, but pixels beyond the plot clamp to the edge rows — for gestures (brush
+ *  releases) that legitimately end outside the plot area. */
+export function rowAtYClamped(layout: HeatmapLayout, y: number): number {
+    if (layout.rows === 0 || layout.rowHeight <= 0) {
+        return -1
+    }
+    return Math.max(0, Math.min(layout.rows - 1, rawRowAtY(layout, y)))
+}
```
[the remark this answers is not in the document]
Comment (jonmcwest) [reply packages/quill/packages/charts/src/charts/Heatmap/heatmap-layout.ts:67]: Good catch — fixed in 7bd102b. `handleAreaSelect` now insets each brush edge by half a pixel (`BRUSH_EDGE_EPSILON_PX`) before snapping to rows, so an edge that lands exactly on a row boundary no longer floors into the neighbouring row and reports a row the rect only touches. The `FULL_HEIGHT_MIN_Y_DRAG_PX` (8px) minimum keeps the span far thicker than the 0.5px nudge, so it can't invert.

<sub>🦉 via [talyn.dev](https://talyn.dev)</sub>

File: packages/quill/packages/charts/src/core/hooks/useChartInteraction.ts [hunk 2]
```diff
@@ -189,13 +194,22 @@ export function useChartInteraction<Meta = unknown>({
         [labels, scales, labelToCoord]
     )
 
+    // Bind the committed scales into the 2D-brush callback so chart-type adapters can map the
+    // y pixel range onto their own bands (the core has no y-band concept).
+    const onAreaSelectWithScales = useMemo(
+        () =>
+            onAreaSelect && scales ? (data: AreaSelectData): void => onAreaSelect(data, scales) : undefined,
+        [onAreaSelect, scales]
+    )
```
[the remark this answers is not in the document]
Comment (jonmcwest) [reply packages/quill/packages/charts/src/core/hooks/useChartInteraction.ts:210]: This is a pre-existing property of the drag gesture rather than something `onAreaSelect` introduces — the same `completeDrag` path in `useDragToZoom` already backs `onDateRangeZoom`, which likewise captures the origin against the mousedown wrapper rect and reads the mouseup coords against the current one. A resize mid-drag is a rare transient, and its only effect is a single, slightly-off selection that self-corrects on the next drag. Re-projecting the origin on every resize would add real complexity for negligible benefit, so I'm leaving the 2D brush consistent with the existing zoom gesture here.

<sub>🦉 via [talyn.dev](https://talyn.dev)</sub>

**Comment by jonmcwest** [comment]
## 🧹 Prepare-PR pass (polish + screenshots)

### Simplify — 1 fix applied
Extracted a shared **`LabelRange`** base type so `DateRangeZoomData` and `AreaSelectData` stop duplicating the four `startLabel`/`endLabel`/`startIndex`/`endIndex` fields and stay in lockstep. Pushed as [`56e6f16`](https://github.com/PostHog/posthog/commit/56e6f16d89e5f5d2d7933dead0a30f4592393050) (type-only, no runtime change).

Everything else the cleanup pass looked at was already at a reasonable floor and left untouched: the x/y clamp "mirror" in `composeDrawHoverWithSelection`, `rowAtYClamped`'s defensive `rows === 0` guard, and the brush-edge epsilon staying in the `Heatmap` adapter (it's brush-specific policy, correctly out of the generic layout helper).

### Code review (high, recall-biased) — no code changes
No confirmed correctness bugs. Two candidates surfaced and were run down:

- **Near-vertical brush is silently dropped.** `completeDrag` gates both callbacks on `dragRectToLabelRange`, whose `SINGLE_LABEL_MIN_DRAG_PX` (8px) guard rejects a drag that stays within one column with <8px of horizontal travel — so a tall, ~1-column-wide brush fires nothing. This is the *documented, intended* reuse of the x min-drag guard ("same `dragRectToLabelRange` resolution and min-drag guard"), and relaxing it would reopen the accidental-drag noise the guard exists to suppress. Flagging as a **design note, not a fix**: if selecting a duration range within a single time bucket is ever a target gesture, the area-select branch would need its own vertical-extent guard rather than borrowing the x one.
- **`wrapper == null` in the global `mouseup` handler fabricates `mouseY = 0`.** Verified **not a regression**: the listener is added/removed by the `zoomEnabled` effect, so during an active gesture the component is mounted and `wrapperRef.current` is set — the null branch is unreachable in practice and only mirrors the pre-existing `mouseX` fallback.

The NaN-in-renderer and "type alias loses declaration merging" candidates were refuted (the renderer path yields `0`, not `NaN`; no consumer merges these payload types).

### Screenshots — skipped (already present)
The change has a visible surface, but the PR description already embeds the `BrushSelection` story mid-drag, and the one polish change is type-only so it renders identically — no recapture needed.

### Verification
This environment has no installed dependencies (`node_modules`/`hogli` absent), so `tsc`/jest couldn't run locally. The `LabelRange` change is type-only and every usage was traced to a plain data-shape position; the pushed commit re-runs the full typecheck and the quill-charts jest suite in CI.

**Comment by jonmcwest** [comment]
Heads up: the initial run of `Visual regression tests - chromium (5/16)` failed, but the failure was a flaky play-test timeout in `products/customer_analytics/frontend/components/Accounts/AccountsTab.stories.tsx` — the `RowExpanded*` interaction stories exceeded the 60s test timeout (retried 3×) while rendering the full `<App/>`. That story is unrelated to this PR, which only touches `packages/quill/packages/charts/`.

I re-ran the failed shard and it passed cleanly, so the failure was transient (CI resource contention on a heavy full-app play-test, not a snapshot regression). All checks are now green.

<sub>🦉 via [talyn.dev](https://talyn.dev)</sub>
